src/Data/Primitive/Contiguous/Class.hs

```diff
@@ -1250,6 +1259,20 @@ instance Contiguous (UnliftedArray_ unlifted_a) where
     unsafeFreezeUnliftedArray m
   {-# INLINE run #-}
   run = runUnliftedArrayST
+  {-# INLINE shrink #-}
+  shrink !arr !n = do
+    -- See Note [Shrinking Unlifted Arrays]
+    cloneMutableUnliftedArray arr 0 n
+  {-# INLINE unsafeShrinkAndFreeze #-}
+  unsafeShrinkAndFreeze !arr !n =
+    -- See Note [Shrinking Unlifted Arrays]
+    freezeUnliftedArray arr 0 n
+
+-- Note [Shrinking Unlifted Arrays]
+-- ================================
+-- This implementation copies the array rather than freezing it in place.
+-- But at least it is able to avoid assigning all of the elements to
+-- a nonsense value before replacing them with memcpy.
```

Comment:
Wow, very clever! But I think you want to apply this to lifted arrays also, right? Including defining a `shrink` using `cloneMutableArray`, and also in Shim.hs changing `resizeArray` to be like `resizeUnliftedArray`.

Reply:
Good idea. Added this change in 95e48d02855bcd8eaf7e55d4340f3f138fc7a35e.